File: package/bin/importers/stig_zip.py
"""Walk DISA STIG product zips and library zip-of-zips for Manual XCCDF baselines."""

from __future__ import annotations

import io
import zipfile
from typing import Any, Dict, Iterator, List, Optional, Tuple, Union

ZIP_MAGIC = b"PK"
MAX_DEPTH = 4
MAX_MEMBER_BYTES = 150 * 1024 * 1024
MAX_BASELINES = 800

ZipSource = Union[bytes, bytearray, str]

_SKIP_SUBSTR = (
    "/__macosx/",
    "_srg_",
    "-srg-",
    "/srg_",
    "srg_v",
    "_scap_",
    "-scap-",
    "/scap/",
    "_ocil",
    "-ocil",
)


def looks_like_zip(data: bytes) -> bool:
    return bool(data) and data[:2] == ZIP_MAGIC


def _norm(name: str) -> str:
    return (name or "").replace("\\", "/").lower()


def is_baseline_xccdf_name(name: str) -> bool:
    """True for DISA Manual XCCDF STIG benchmarks, not SRGs/SCAP/checklists."""
    path = _norm(name)
    base = path.rsplit("/", 1)[-1]
    if not base.endswith(".xml") or base.startswith("."):
        return False
    if "manual-xccdf" not in base:
        return False
    if any(part in path for part in _SKIP_SUBSTR):
        return False
    return True


def skip_reason(name: str) -> str:
    path = _norm(name)
    if is_baseline_xccdf_name(name):
        return ""
    if "manual-xccdf" in path and any(
        part in path for part in ("_srg_", "-srg-", "/srg_", "srg_v")
    ):
        return "srg"
    if "scap" in path or "ocil" in path:
        return "scap"
    return "other"


def _open_zip(source: ZipSource) -> zipfile.ZipFile:
    if isinstance(source, (bytes, bytearray)):
        return zipfile.ZipFile(io.BytesIO(source))
    return zipfile.ZipFile(source)


def iter_baseline_xccdfs(
    data: bytes,
    *,
    source_prefix: str = "",
    depth: int = 0,
    _count: Optional[list] = None,
) -> Iterator[Tuple[str, bytes]]:
    yield from iter_baseline_xccdfs_from(
        data, source_prefix=source_prefix, depth=depth, _count=_count
    )


def iter_baseline_xccdfs_from(
    source: ZipSource,
    *,
    source_prefix: str = "",
    depth: int = 0,
    _count: Optional[list] = None,
) -> Iterator[Tuple[str, bytes]]:
    if depth > MAX_DEPTH:
        return
    count = _count if _count is not None else [0]
    try:
        archive = _open_zip(source)
    except (zipfile.BadZipFile, OSError) as err:
        raise ValueError(f"invalid zip: {err}") from err
    with archive:
        for info in archive.infolist():
            if info.is_dir() or info.file_size <= 0:
                continue
            member = (source_prefix + info.filename).replace("\\", "/")
            if info.file_size > MAX_MEMBER_BYTES:
                continue
            try:
                raw = archive.read(info)
            except Exception:
                continue
            if looks_like_zip(raw) and depth < MAX_DEPTH:
                yield from iter_baseline_xccdfs_from(
                    raw,
                    source_prefix=member.rstrip("/") + "/",
                    depth=depth + 1,
                    _count=count,
                )
                continue
            if is_baseline_xccdf_name(member):
                count[0] += 1
                if count[0] > MAX_BASELINES:
                    raise ValueError(
                        f"zip contains more than {MAX_BASELINES} Manual-xccdf baselines"
                    )
                yield member, raw


def list_baseline_xccdfs(data: bytes) -> List[Tuple[str, bytes]]:
    found = list(iter_baseline_xccdfs_from(data))
    if not found:
        raise ValueError(
            "no Manual-xccdf STIG baselines found (SRGs, SCAP, and checklists are skipped)"
        )
    return found


def explore_baseline_zip(source: ZipSource) -> Dict[str, Any]:
    """List Manual-xccdf members without loading every XML body."""
    found: List[Dict[str, Any]] = []
    skipped = {"srg": 0, "scap": 0, "other": 0}

    def walk(src: ZipSource, prefix: str, depth: int) -> None:
        if depth > MAX_DEPTH:
            return
        archive = _open_zip(src)
        with archive:
            for info in archive.infolist():
                if info.is_dir() or info.file_size <= 0:
                    continue
                member = (prefix + info.filename).replace("\\", "/")
                nested = member.lower().endswith(".zip") and depth < MAX_DEPTH
                if nested:
                    if info.file_size > MAX_MEMBER_BYTES:
                        skipped["other"] += 1
                        continue
                    try:
                        raw = archive.read(info)
                    except Exception:
                        skipped["other"] += 1
                        continue
                    if looks_like_zip(raw):
                        walk(raw, member.rstrip("/") + "/", depth + 1)
                    continue
                if is_baseline_xccdf_name(member):
                    if len(found) >= MAX_BASELINES:
                        raise ValueError(
                            f"zip contains more than {MAX_BASELINES} Manual-xccdf baselines"
                        )
                    found.append({"path": member, "size": int(info.file_size)})
                    continue
                reason = skip_reason(member)
                skipped[reason] = skipped.get(reason, 0) + 1

    walk(source, "", 0)
    if not found:
        raise ValueError(
            "no Manual-xccdf STIG baselines found (SRGs, SCAP, and checklists are skipped)"
        )
    return {"found": found, "skipped": skipped}


def extract_baseline_xccdf(source: ZipSource, want_path: str) -> bytes:
    want = (want_path or "").replace("\\", "/")
    if not is_baseline_xccdf_name(want):
        raise ValueError("not a Manual-xccdf baseline path")

    def walk(src: ZipSource, prefix: str, depth: int) -> Optional[bytes]:
        if depth > MAX_DEPTH:
            return None
        archive = _open_zip(src)
        with archive:
            for info in archive.infolist():
                if info.is_dir() or info.file_size <= 0:
                    continue
                member = (prefix + info.filename).replace("\\", "/")
                nested = member.lower().endswith(".zip") and depth < MAX_DEPTH
                if nested:
                    if info.file_size > MAX_MEMBER_BYTES:
                        continue
                    try:
                        raw = archive.read(info)
                    except Exception:
                        continue
                    if looks_like_zip(raw):
                        hit = walk(raw, member.rstrip("/") + "/", depth + 1)
                        if hit is not None:
                            return hit
                    continue
                if member == want:
                    return archive.read(info)
        return None

    data = walk(source, "", 0)
    if not data:
        raise ValueError(f"member not found: {want}")
    return data

File: ui/src/stigZip.js
import {
    BlobReader,
    BlobWriter,
    Uint8ArrayWriter,
    ZipReader,
    configure,
} from "@zip.js/zip.js";

configure({ useWebWorkers: false });

const MAX_DEPTH = 4;
const MAX_XCCDF_BYTES = 50 * 1024 * 1024;
const MAX_INNER_ZIP_BYTES = 150 * 1024 * 1024;
const MAX_BASELINES = 500;

const SKIP = [
    "/__macosx/",
    "_srg_",
    "-srg-",
    "/srg_",
    "srg_v",
    "_scap_",
    "-scap-",
    "/scap/",
    "_ocil",
    "-ocil",
];

function norm(name) {
    return String(name || "").replace(/\\/g, "/");
}

export function isZipName(name) {
    return norm(name).toLowerCase().endsWith(".zip");
}

export function isBaselineXccdfName(name) {
    const path = norm(name).toLowerCase();
    const base = path.split("/").pop() || "";
    if (!base.endsWith(".xml") || base.startsWith(".")) {
        return false;
    }
    if (base.indexOf("manual-xccdf") < 0) {
        return false;
    }
    return !SKIP.some((part) => path.indexOf(part) >= 0);
}

export function basename(name) {
    const path = norm(name);
    const parts = path.split("/").filter(Boolean);
    return parts[parts.length - 1] || path;
}

async function withZip(blob, fn) {
    const reader = new ZipReader(new BlobReader(blob), { useWebWorkers: false });
    try {
        return await fn(await reader.getEntries());
    } finally {
        await reader.close();
    }
}

function decodeXml(bytes) {
    const data = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes || []);
    if (data.length >= 2 && data[0] === 0xff && data[1] === 0xfe) {
        return new TextDecoder("utf-16le").decode(data);
    }
    if (data.length >= 2 && data[0] === 0xfe && data[1] === 0xff) {
        return new TextDecoder("utf-16be").decode(data);
    }
    return new TextDecoder("utf-8").decode(data);
}

function skipReason(name) {
    const path = norm(name).toLowerCase();
    if (SKIP.some((part) => path.indexOf(part) >= 0) && path.indexOf("manual-xccdf") >= 0) {
        return "srg";
    }
    if (path.indexOf("scap") >= 0 || path.indexOf("ocil") >= 0) {
        return "scap";
    }
    if (path.endsWith(".xml") && path.indexOf("manual-xccdf") >= 0) {
        return "srg";
    }
    return "other";
}

async function walk(blob, prefix, depth, handlers, mode) {
    if (depth > MAX_DEPTH) {
        return;
    }
    await withZip(blob, async (entries) => {
        for (let i = 0; i < entries.length; i += 1) {
            const entry = entries[i];
            if (entry.directory) {
                continue;
            }
            const member = norm(prefix + entry.filename);
            const size = Number(entry.uncompressedSize || 0);
            if (isZipName(member) && depth < MAX_DEPTH) {
                if (size > MAX_INNER_ZIP_BYTES) {
                    continue;
                }
                if (handlers.onProgress) {
                    handlers.onProgress({
                        phase: "scan",
                        zip: basename(member),
                        found: handlers.found || 0,
                    });
                }
                const inner = await entry.getData(new BlobWriter("application/zip"));
                await walk(inner, member.replace(/\/?$/, "/") , depth + 1, handlers, mode);
                continue;
            }
            if (!isBaselineXccdfName(member)) {
                if (mode === "list" && handlers.onSkip) {
                    handlers.onSkip(member, skipReason(member));
                }
                continue;
            }
            if (handlers.only && !handlers.only.has(member)) {
                continue;
            }
            if (mode === "list") {
                handlers.paths.push({ path: member, size });
                handlers.found = handlers.paths.length;
                if (handlers.paths.length > MAX_BASELINES) {
                    throw new Error(
                        "zip contains more than " + MAX_BASELINES + " Manual-xccdf baselines"
                    );
                }
                if (handlers.onProgress) {
                    handlers.onProgress({
                        phase: "scan",
                        found: handlers.paths.length,
                        path: member,
                    });
                }
                continue;
            }
            if (size > MAX_XCCDF_BYTES) {
                await handlers.onEach(member, null, "file too large");
                continue;
            }
            const raw = await entry.getData(new Uint8ArrayWriter());
            await handlers.onEach(member, decodeXml(raw), "");
        }
    });
}

export async function listBaselineXccdfs(blob, onProgress) {
    const explored = await exploreStigZip(blob, onProgress);
    return explored.found.map((item) => item.path);
}

export async function exploreStigZip(blob, onProgress) {
    const skipped = { srg: 0, scap: 0, other: 0 };
    const handlers = {
        paths: [],
        found: 0,
        onProgress,
        onSkip: (name, reason) => {
            if (skipped[reason] != null) {
                skipped[reason] += 1;
            } else {
                skipped.other += 1;
            }
        },
    };
    await walk(blob, "", 0, handlers, "list");
    if (!handlers.paths.length) {
        throw new Error(
            "no Manual-xccdf STIG baselines found (SRGs, SCAP, and checklists are skipped)"
        );
    }
    return { found: handlers.paths, skipped };
}

export async function forEachBaselineXccdf(blob, paths, onEach, onProgress) {
    const only = new Set(paths || []);
    const handlers = {
        only,
        found: only.size,
        onProgress,
        onEach: async (path, xml, error) => {
            if (onProgress) {
                onProgress({ phase: "read", path });
            }
            await onEach(path, xml, error);
        },
    };
    await walk(blob, "", 0, handlers, "extract");
}
